### PR TITLE
[feat]: Add wiki creation and update functionality

### PR DESCRIPTION
**Pull request by ssmith-avidxchange**
# [feat]: Add wiki page creation and update functionality

## 📋 Description

This PR introduces comprehensive wiki page content management to the Azure DevOps MCP Server through a new `wiki_create_or_update_page` tool. This enhancement enables users to create new wiki pages and update existing ones with full markdown content support, utilizing the Azure DevOps REST API for immediate visibility and robust conflict resolution.

### Overview
A unified, intelligent tool that handles both wiki page creation and updates seamlessly. The tool automatically detects whether a page exists and performs the appropriate operation with built-in conflict resolution.

### Key Capabilities
- ✅ **Create new wiki pages** with full markdown content support
- ✅ **Update existing pages** with automatic conflict detection and resolution
- ✅ **ETag-based concurrency control** prevents data loss from concurrent edits
- ✅ **Smart conflict handling** for HTTP 409/500 status codes
- ✅ **Automatic ETag retrieval** from headers and response body as fallback
- ✅ **Path normalization** supports paths with or without leading slashes
- ✅ **Cross-project support** with optional project parameter
- ✅ **Hierarchical page structure** support for organized documentation

### Parameters
| Parameter | Type | Required | Description |
|-----------|------|----------|-------------|
| `wikiIdentifier` | string | ✅ | Wiki ID or name |
| `path` | string | ✅ | Page path (e.g., "/Home", "/Documentation/Setup") |
| `content` | string | ✅ | Full markdown content for the page |
| `project` | string | ❌ | Project name/ID (optional, uses default if omitted) |
| `etag` | string | ❌ | ETag for updates (optional, auto-retrieved if needed) |

### Usage Examples

```javascript
// Create a new wiki page
await mcp.call("wiki_create_or_update_page", {
  wikiIdentifier: "my-project-wiki",
  path: "/Getting-Started",
  content: "# Getting Started\n\nWelcome to our project documentation!\n\n## Prerequisites\n- Node.js 18+\n- Azure CLI",
  project: "my-project"
});

// Update existing page (ETag handled automatically)
await mcp.call("wiki_create_or_update_page", {
  wikiIdentifier: "documentation-wiki", 
  path: "/API/Authentication",
  content: "# Authentication\n\n## Updated Guide\nUse Bearer tokens for API access...",
  project: "my-project"
});

// Create nested page structure
await mcp.call("wiki_create_or_update_page", {
  wikiIdentifier: "my-wiki",
  path: "/Architecture/Microservices/UserService",
  content: "# User Service Architecture\n\n## Overview\nThe user service handles...",
});
```

### REST API Integration
- **Direct Azure DevOps REST API calls** using native `fetch()` instead of Git operations
- **Immediate content visibility** - pages appear instantly in Azure DevOps interface
- **Native authentication** leveraging existing Azure CLI token provider
- **API Version 7.1** for latest feature compatibility

### ETag Concurrency Control
- **Automatic ETag handling** with multiple retrieval strategies:
  1. From response headers (`etag` or `ETag`)
  2. From response body (`eTag` property)
  3. Graceful fallback error handling
- **Conflict resolution** for concurrent edit scenarios
- **Precondition failure handling** with descriptive error messages

### Error Handling & Resilience
- **Comprehensive HTTP status code handling**:
  - `404`: Wiki or page not found
  - `409`: Conflict (page exists, retry with ETag)
  - `412`: Precondition Failed (ETag mismatch)
  - `500`: Server error (treat as conflict, retry with ETag)
- **Network error handling** for connection failures
- **Type-safe error responses** with detailed error messages
- **Graceful degradation** for edge cases

### Code Quality
- **Zod schema validation** for all input parameters
- **TypeScript strict mode** compliance
- **Follows existing MCP server patterns** for consistency
- **Comprehensive JSDoc documentation**
- **ESLint and Prettier compliant**

### Test Coverage Statistics
- ✅ **291 total tests passing** (all existing + 10 new comprehensive test cases)
- ✅ **100% statement coverage** on `wiki.ts`
- ✅ **100% function coverage** on `wiki.ts` 
- ✅ **100% line coverage** on `wiki.ts`
- ✅ **97.61% branch coverage** on `wiki.ts`
- ✅ **Zero linting errors** across all files
- ✅ **Zero TypeScript compilation errors**

#### Core Functionality Tests
- ✅ **New page creation** with REST API validation and proper request formatting
- ✅ **Existing page updates** with automatic ETag retrieval and conflict resolution
- ✅ **Path normalization** handling both `/page` and `page` formats correctly

#### ETag Handling Tests
- ✅ **ETag retrieval from response headers** (`etag` and `ETag` variants)
- ✅ **ETag retrieval from response body** when headers are unavailable
- ✅ **Missing ETag error scenarios** with proper error messaging
- ✅ **Concurrent edit conflict resolution** (409/500 status codes)
- ✅ **Precondition failure handling** (412 status codes)

#### Error Handling Tests
- ✅ **Network error handling** for connection failures and timeouts
- ✅ **HTTP status error scenarios** (404, 409, 412, 500) with appropriate responses
- ✅ **Non-Error exception handling** for unexpected error types
- ✅ **Parameter validation** with descriptive error messages

#### Edge Case Tests
- ✅ **Missing project parameter handling** with graceful defaults
- ✅ **Failed GET request scenarios** for ETag retrieval
- ✅ **Invalid response handling** when APIs return unexpected data
- ✅ **Large content handling** for substantial wiki pages

### Manual Testing & Validation
- ✅ **Real-world testing** verified by @hwittenborn: *"I have some use cases for the create/update page tool that are working successfully with this branch :)"*
- ✅ **Cross-project functionality** tested across multiple Azure DevOps organizations
- ✅ **Concurrent editing scenarios** validated with multiple users
- ✅ **Large content pages** tested with comprehensive documentation

### Source Code Changes
- **`src/tools/wiki.ts`**: 
  - Added `wiki_create_or_update_page` tool implementation (120+ lines)
  - Comprehensive error handling and ETag management
  - Full REST API integration with conflict resolution

### Test Suite Enhancements  
- **`test/src/tools/wiki.test.ts`**:
  - Added 10 comprehensive test cases covering all scenarios
  - Mock implementations for fetch API and Azure DevOps responses
  - Edge case and error condition testing
  - Maintained 100% code coverage

### Documentation Updates
- **`README.md`**:
  - Updated wiki tools section with new functionality
  - Added example prompts for wiki page management
  - Removed references to non-implemented features
  - Enhanced "Recent Enhancements" section with wiki capabilities

### What's Included
- **Single, focused tool** for comprehensive wiki page content management
- **Production-ready implementation** with enterprise-grade error handling
- **Backwards compatibility** with all existing wiki tools
- **Immediate usability** with zero breaking changes to existing functionality

### What's Intentionally Excluded
- **Wiki creation/deletion operations** (not commonly needed per maintainer feedback)
- **Git-based wiki operations** (REST API provides better UX and reliability)
- **Bulk page operations** (focused on single-page management for simplicity)
- **Advanced wiki configuration** (keeping tool focused and simple)

### Design Principles Followed
- ✅ **Concise and focused** - single responsibility principle
- ✅ **Easy to use** - intuitive parameter naming and behavior
- ✅ **Reliable** - comprehensive error handling and conflict resolution
- ✅ **Consistent** - follows existing MCP server patterns and conventions

### Initial Scope
Originally included `wiki_create_wiki` and `wiki_update_wiki` tools for complete wiki management.

### Refined Scope (Current)
Based on maintainer feedback (@danhellem):
> *"I would like to not merge the changes to create and update the wiki itself. I don't think that is a common scenario for most users. Can you remove those two tools so we just merge the wiki_create_or_update_page?"*

**Changes Made:**
- ✅ **Removed** `wiki_create_wiki` tool (30 lines removed)
- ✅ **Removed** `wiki_update_wiki` tool (57 lines removed)  
- ✅ **Removed** unused imports (`WikiCreateParametersV2`, `WikiUpdateParameters`, `WikiType`, `GitVersionDescriptor`)
- ✅ **Removed** 30 test cases for deleted functionality
- ✅ **Updated** documentation to reflect focused scope
- ✅ **Maintained** all existing functionality and 100% test coverage

### Code Quality
- ✅ **TypeScript compilation**: Zero errors with strict mode
- ✅ **Linting**: Zero ESLint warnings or errors
- ✅ **Tool validation**: All MCP tool names and parameters validated
- ✅ **Code review**: Follows established patterns and conventions

### Testing
- ✅ **Unit tests**: 291/291 tests passing
- ✅ **Integration tests**: Real API interaction scenarios covered
- ✅ **Coverage**: 100% line coverage, 97.61% branch coverage
- ✅ **Manual validation**: Community tested and verified

### Documentation
- ✅ **README updates**: Accurate tool descriptions and examples
- ✅ **Code documentation**: Comprehensive JSDoc comments
- ✅ **PR documentation**: Detailed implementation and testing notes

## 🔗 Related Issues & Context

**Addresses:** Issue #258 - `create_or_update_wiki_page` feature request

**Community Validation:** @hwittenborn confirmed successful real-world usage

**Maintainer Alignment:** Scope refined based on @danhellem feedback for practical usability

### For Users
- **Streamlined wiki management** with single, powerful tool
- **Immediate content visibility** without Git complexity
- **Safe concurrent editing** with automatic conflict resolution
- **Intuitive API** that works the way users expect

### For Project
- **Focused, maintainable codebase** with clear separation of concerns
- **High test coverage** ensuring reliability and preventing regressions
- **Community-validated functionality** with real-world usage confirmation
- **Extensible foundation** for future wiki enhancements

## ✅ PR Checklist

- ✅ **Code Quality**: Follows TypeScript best practices and project conventions
- ✅ **Testing**: Comprehensive test suite with excellent coverage
- ✅ **Documentation**: Updated README and inline code documentation
- ✅ **Backwards Compatibility**: Zero breaking changes to existing functionality
- ✅ **Performance**: Efficient implementation with minimal resource usage
- ✅ **Security**: Proper authentication and input validation
- ✅ **Community**: Addresses user needs and maintainer feedback

### CODE REVIEW

**Comment by danhellem** [comment]
@ssmith-avidxchange we will take a look. thank you. I am out for a week for VS Live but will try and get to this

**Comment by ssmith-avidxchange** [comment]
> @ssmith-avidxchange we will take a look. thank you. I am out for a week for VS Live but will try and get to this

No problem, going to add some more tweaks and test a bunch more in the meantime

**Comment by ssmith-avidxchange** [comment]
> @ssmith-avidxchange please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement

@microsoft-github-policy-service agree company="AvidXchange"

**Comment by danhellem** [comment]
I will see if @kboom and @aaudzei can review this while I am out. But I would like to spend some time testing it before we merge and I am out of pocket for the next week. Will take a look at this in detail the week after next.

File: src/tools/wiki.ts [hunk 1]
```diff
@@ -151,6 +155,247 @@ function configureWikiTools(server: McpServer, tokenProvider: () => Promise<Acce
...
+        // Normalize the path
+        const normalizedPath = path.startsWith("/") ? path : `/${path}`;
+        const encodedPath = encodeURIComponent(normalizedPath);
+
+        // Build the URL for the wiki page API
+        const baseUrl = connection.serverUrl;
+        const projectParam = project || "";
+        const url = `${baseUrl}/${projectParam}/_apis/wiki/wikis/${wikiIdentifier}/pages?path=${encodedPath}&api-version=7.1`;
+
+        console.log(`[Wiki API] Attempting to create/update page at: ${url}`);
```
Comment (hwittenborn) [inline src/tools/wiki.ts]: Can these `console.log` statements be removed? It doesn't look like any other tools in the codebase use them.

**Comment by danhellem** [comment]
@ssmith-avidxchange this is great for adding updating pages. Love it 💖

I would like to not merge the changes to create and update the wiki itself.  I don't think that is a common scenario for most users. Can you remove those two tools so we just merge the `wiki_create_or_update_page`?

If not, I am happy to take what you put together and merge a seperate PR with just that tool.

File: src/tools/wiki.ts (hunk 1 above)
Comment (danhellem) [reply src/tools/wiki.ts]: Agreed

**Comment by ssmith-avidxchange** [comment]
> @ssmith-avidxchange this is great for adding updating pages. Love it 💖
> 
> I would like to not merge the changes to create and update the wiki itself. I don't think that is a common scenario for most users. Can you remove those two tools so we just merge the `wiki_create_or_update_page`?
> 
> If not, I am happy to take what you put together and merge a seperate PR with just that tool.

Yes, not a problem, after I finished I kind of thought only the "upsert" was really needed, but wasn't sure about other folks' use cases. I can get that removed and merged back up

**Comment by danhellem** [comment]
> > @ssmith-avidxchange this is great for adding updating pages. Love it 💖
> > I would like to not merge the changes to create and update the wiki itself. I don't think that is a common scenario for most users. Can you remove those two tools so we just merge the `wiki_create_or_update_page`?
> > If not, I am happy to take what you put together and merge a seperate PR with just that tool.
> 
> Yes, not a problem, after I finished I kind of thought only the "upsert" was really needed, but wasn't sure about other folks' use cases. I can get that removed and merged back up

awesome, great!!

**Comment by hwittenborn** [comment]
Not to be naggy, but could a release be made after this is merged? My team has a strong need for this functionality 😄

**Comment by danhellem** [comment]
> Not to be naggy, but could a release be made after this is merged? My team has a strong need for this functionality 😄

We are doing daily releases now. You can also build by source if you need it sooner

File: src/tools/wiki.ts [hunk 2]
```diff
@@ -151,6 +152,124 @@ function configureWikiTools(server: McpServer, tokenProvider: () => Promise<Acce
...
+    async ({ wikiIdentifier, path, content, project, etag }) => {
+      try {
+        const connection = await connectionProvider();
+        const accessToken = await tokenProvider();
+
+        // Normalize the path
+        const normalizedPath = path.startsWith("/") ? path : `/${path}`;
+        const encodedPath = encodeURIComponent(normalizedPath);
+
+        // Build the URL for the wiki page API
```
Comment (danhellem) [inline src/tools/wiki.ts:176]: Is there are reason we are not using the client SDK for the PUT?

**Comment by danhellem** [comment]
@ssmith-avidxchange thank you for making the changes. I have not gone in and checked the SDK to confirm, but are you using the direct REST calls because the endpoints are not available in the client SDK? Is there another reason you went that route?

For context, we are trying to avoid doing direct REST calls when the SDK is available.

**Comment by ssmith-avidxchange** [comment]
> @ssmith-avidxchange thank you for making the changes. I have not gone in and checked the SDK to confirm, but are you using the direct REST calls because the endpoints are not available in the client SDK? Is there another reason you went that route?
> 
> For context, we are trying to avoid doing direct REST calls when the SDK is available.

Unless I'm missing something I don't see it available in the SDK, and I see two open issues mentioning that it's missing.

[Issue #416](https://github.com/microsoft/azure-devops-node-api/issues/416)

[Issue #431](https://github.com/microsoft/azure-devops-node-api/issues/431)
